examples/research_projects/jax-projects/dataset-streaming/run_mlm_flax_stream.py

```diff
@@ -0,0 +1,652 @@
+
+    # Define gradient update step fn
+    def train_step(state, batch, dropout_rng):
+        dropout_rng, new_dropout_rng = jax.random.split(dropout_rng)
+
+        def loss_fn(params):
+            labels = batch.pop("labels")
+
+            logits = state.apply_fn(**batch, params=params, dropout_rng=dropout_rng, train=True)[0]
+
+            # compute loss, ignore padded input tokens
+            label_mask = jnp.where(labels > 0, 1.0, 0.0)
+            loss = optax.softmax_cross_entropy(logits, onehot(labels, logits.shape[-1])) * label_mask
+
+            # take average
+            loss = loss.sum() / label_mask.sum()
+
+            return loss
+
+        grad_fn = jax.value_and_grad(loss_fn)
+        loss, grad = grad_fn(state.params)
+        grad = jax.lax.pmean(grad, "batch")
+        new_state = state.apply_gradients(grads=grad)
+
+        metrics = jax.lax.pmean(
+            {"loss": loss, "learning_rate": linear_decay_lr_schedule_fn(state.step)}, axis_name="batch"
+        )
+
+        return new_state, metrics, new_dropout_rng
+
+    # Create parallel version of the train step
+    p_train_step = jax.pmap(train_step, "batch", donate_argnums=(0,))
+
+    # Define eval fn
+    def eval_step(params, batch):
+        labels = batch.pop("labels")
+
+        logits = model(**batch, params=params, train=False)[0]
+
+        # compute loss, ignore padded input tokens
+        label_mask = jnp.where(labels > 0, 1.0, 0.0)
+        loss = optax.softmax_cross_entropy(logits, onehot(labels, logits.shape[-1])) * label_mask
+
+        # compute accuracy
+        accuracy = jnp.equal(jnp.argmax(logits, axis=-1), labels) * label_mask
+
+        # summarize metrics
+        metrics = {"loss": loss.sum(), "accuracy": accuracy.sum(), "normalizer": label_mask.sum()}
+        metrics = jax.lax.psum(metrics, axis_name="batch")
+
+        return metrics
+
+    p_eval_step = jax.pmap(eval_step, "batch", donate_argnums=(0,))
+
+    # Replicate the train state on each device
+    state = jax_utils.replicate(state)
+
+    train_time = 0
+    #    epochs = tqdm(range(num_epochs), desc=f"Epoch ... (1/{num_epochs})", position=0)
+    num_train_samples_per_streaming = num_steps_per_stream_epoch * train_batch_size
+    stream_epoch = 0
+    next_stream_batch = None
+
+    # ======================== Training ================================
+    for step in range(num_train_steps):
+        if step % num_steps_per_stream_epoch == 0:
+            stream_epoch += 1
+            train_start = time.time()
+            train_metrics = []
+
+            # Create sampling rng
+            rng, input_rng = jax.random.split(rng)
+
+            next_stream_batch = [next(iter(tokenized_datasets)) for _ in range(num_train_samples_per_streaming)]
+
+        # Generate an stream epoch by shuffling sampling indices from the train dataset
+        train_samples_idx = jax.random.permutation(input_rng, jnp.arange(num_train_samples_per_streaming))
```

Comment:
I shuffle within the streaming batch - think this should be good enough actually assuming that the streaming batch is big enough. Should we maybe shuffle the dataset right in the beginning after it's "stream" loaded as you explained here: https://huggingface.co/docs/datasets/dataset_streaming.html?highlight=streaming#shuffle-the-dataset @lhoestq ?

Reply:
Can be deleted here - and use official shuffle instead